pkg/syncer/metadatasyncer.go

```diff
@@ -855,6 +857,89 @@ func addLabelsToTopologyVCMap(ctx context.Context, nodeTopoObj csinodetopologyv1
 	}
 }
 
+// cnsvolumeoperationrequestCRAdded checks if the cnsvolumeoperationrequest instance reserved field is present
+// and updates the reserved field for StoragePolicUsage CR
+func cnsvolumeoperationrequestCRAdded(obj interface{}) {
+	ctx, log := logger.GetNewContextWithLogger()
+	// Verify objects received.
+	var (
+		cnsvolumeoperationrequestObj cnsvolumeoperationrequestv1alpha1.CnsVolumeOperationRequest
+	)
+	err := runtime.DefaultUnstructuredConverter.FromUnstructured(
+		obj.(*unstructured.Unstructured).Object, &cnsvolumeoperationrequestObj)
+	if err != nil {
+		log.Errorf("cnsvolumeoperationrequestCRAdded: failed to cast object %+v to %s. Error: %+v", obj,
+			cnsvolumeoperationrequest.CRDSingular, err)
+		return
+	}
+	// Check for the below set of conditions:
+	// 1. Cnsvolumeoperationrequest object's StorageQuotaDetails should not be nil
+	// 2. Cnsvolumeoperationrequest object's StorageQuotaDetails.Reserved should not be nil
+	if cnsvolumeoperationrequestObj.Status.StorageQuotaDetails != nil &&
+		cnsvolumeoperationrequestObj.Status.StorageQuotaDetails.Reserved != nil {
+		restConfig, err := config.GetConfig()
+		if err != nil {
+			log.Errorf("cnsvolumeoperationrequestCRAdded: failed to get Kubernetes config. Err: %+v", err)
+			return
+		}
+		cnsOperatorClient, err := k8s.NewClientForGroup(ctx, restConfig, cnsoperatorv1alpha1.GroupName)
+		if err != nil {
+			log.Errorf("cnsvolumeoperationrequestCRAdded: Failed to create CnsOperator client. Err: %+v", err)
+			return
+		}
+		namespace := cnsvolumeoperationrequestObj.Status.StorageQuotaDetails.Namespace
+		storagePolicyUsageList := &storagepolicyusagev1alpha1.StoragePolicyUsageList{}
+		err = cnsOperatorClient.List(ctx, storagePolicyUsageList, client.InNamespace(namespace))
+		if err != nil {
+			log.Errorf("failed to list %s CR from supervisor namespace %q. Error: %+v",
+				storagepolicyusagev1alpha1.CRDSingular, namespace, err)
+			return
+		}
+		foundStoragePolicyUsageCR := false
+		oldStoragePolicyUsageCR := &storagepolicyusagev1alpha1.StoragePolicyUsage{}
+		newStoragePolicyUsageCR := &storagepolicyusagev1alpha1.StoragePolicyUsage{}
+		if len(storagePolicyUsageList.Items) > 0 {
+			for _, item := range storagePolicyUsageList.Items {
+				policyId := cnsvolumeoperationrequestObj.Status.StorageQuotaDetails.StoragePolicyId
+				scName := cnsvolumeoperationrequestObj.Status.StorageQuotaDetails.StorageClassName
```

Comment:
We will improve this code to directly get the object we need to update instead of traversing the loop.

Reply:
Tracking this enhancement here - https://github.com/kubernetes-sigs/vsphere-csi-driver/issues/2707